### DISCUSSION TITLE
Boundary Conditions - Ghost cell values from previous time step/cycle

### DISCUSSION THREAD

**Discussion by kyriakosTapinou** [PrincetonUniversity/athena#564] [Q&A] [answered]
I need to specify an evolving magnetic field on the domain boundary. If the previous time step / cycle data is available then this become an easy implementation if the numerical update is known. Is there an existing way to access previous cycle ghost cell and face centered values?

**Comment by c-white** [comment, marked as answer]
The values from half a previous substage of the integrator might be lingering around in memory in some cases, but in general the code does not store values once they are no longer needed.

For this purpose, I would suggest allocating your own `ruser_meshblock_data` arrays, [see here](https://github.com/PrincetonUniversity/athena/wiki/Problem-Generators#user-defined-data-field). Then toward the end of the timestep, probably in `MeshBlock::UserWorkInLoop`, copy whatever values you want into your arrays.

**Reply by kyriakosTapinou** [reply]
That will be fine. 

In addition to evolving a boundary property, have you seen the user-defined boundary condition function being linked to an external database or library/program that can feed data in-between cycles. I've done a similar thing in the past with a FEM structural solver and a CFD solver to get a coupled structure-fluid simulation. 

Would this best be accomplished with the UserWorkAfterLoop running the external program and piping desired data to the iuser_meshblock_data, then relying on this in the user boundary condition function?

**Reply by c-white** [reply]
I haven't seen this done, but something like what you describe could probably work. One thing to be careful about is the difference between Mesh functions (run once on each MPI rank) and arrays (unique to each MPI rank, only rank 0 stored in restarts) vs MeshBlock functions (called for each MeshBlock) and arrays (local to each MeshBlock). If your external function should be called once given the global state of the system, the outline for `Mesh::UserWorkInLoop` would be something like
- Gather all relevant data from MeshBlocks on this Mesh
- Gather all data to rank 0
- Call external function on rank 0
- Broadcast data from rank 0 to all ranks
- Possibly distribute data to all MeshBlocks on this Mesh (e.g. to `ruser_meshblock_data` arrays)

Also, UserWorkAfterLoop runs after the code has finished evolving; UserWorkInLoop runs at the end of each timestep.

**Reply by kyriakosTapinou** [reply]
Thanks for your answer, this sounds like a promising avenue. I'll be looking into this further. 

In the user boundary conditions, which ghost cell center and ghost cell face values are initialised to zero and/or overridden? 
When running with more than one block, I see nan's creeping into the domain from ghost cell values in ruser_meshblock_data. This happens even though I initialise the ruser_meshblock_data to zero everywhere before usage. The update is essentially 
1) InitUserMeshData
2) InitUserMeshBlockData - > ruser_meshblock_data init to zero everywhere 
3) MeshBlock ProblemGenerator
4) UserBoundaryFunction - > set ghost cell face values for b fields 
                                          - > b = ruser + update 
The  the cycle 
1) UserBoundaryFunction - > set ghost cell face values for b fields 
                                          - > b = ruser + update 
2) UserWorkInLoop           - > Update ruser = ghost cell face values for b fields 

Nans are creeping in at odd places on block boundaries. Are there specific ghost cells overwritten automatically in the blocking stage at the start of the cycle?

**Reply by c-white** [reply]
The user boundary condition should consider the input `is`, `ie`, etc. parameters to mark the active cells (inclusive on both ends), and should set the first `ngh` cells beyond this in the appropriate direction, as well as the first `ngh` faces in each orientation in the following sense.

Suppose the function is enrolled as `BoundaryFace::outer_x1`. Then cells `ie + 1 <= i <= ie + ngh`, `js <= j <= je`, `ks <= k <= ke` should be set; x-faces `ie + 2 <= i <= ie + ngh + 1`, `js <= j <= je`, `ks <= k <= ke` should be set; y-faces `ie + 1 <= i <= ie + ngh`, `js <= j <= je + 1`, `ks <= k <= ke` should be set; and z-faces `ie + 1 <= i <= ie + ngh`, `js <= j <= je`, `ks <= k <= ke + 1` should be set.

There is a subtle extra complexity that comes from using mesh refinement. In order to enable proper prolongation/restriction, the code has to create virtual buffers near refinement boundaries. If these touch a physical boundary, the buffers may need to have the boundary function called on themselves. This is why the active cell limits and number of ghost cells are parameters given to the function -- they may be very different from the values you expect. If this is the case, the above solution of storing old values will need to be modified.

**Reply by kyriakosTapinou** [reply]
Thanks. Unfortunately there is some subtlety I am missing without refinement. My boundary condition and ruser_meshblock_data register work fine in a single block, but not with multiple blocks. Is there any issue with using kl or ks-NGHOST etc. when moving between single and multiple blocks?

**Reply by tomidakn** [reply]
When mesh refinement is in use, the boundary conditions are also applied on "coarse buffer", which is a data grid overlapping with a MeshBlock but with half resolution. The index ranges are different on this coarse buffer, so make sure you always use the loop limits provided from the boundary function parameters (e.g., use ngh, not NGHOST).

**Reply by kyriakosTapinou** [reply]
@tomidakn sorry I should specify that I am trying to coordinate ruser_meshblock_data updates in UserWorkInLoop and the user boudnary function for the inner x3 boundary. What is the appropriate call to ngh from UserWorkInLoop? I have looked through the docs for relevant descriptions but couldn't find one, if you could point me in the right direction or have some suggestions I'd be grateful. 
Also if there is also a portion of the docs that speaks about definitions of internal block boundaries that would be very useful too. Otherwise is the only good resource the paper "The Athena++ Adaptive Mesh Refinement Framework: Design and Magnetohydrodynamic Solvers"

Thanks for your continued help!

**Reply by c-white** [reply]
If there is *no* mesh refinement, then I'm reasonably sure `ngh` will always be `NGHOST`, and loop limits like `ks-NGHOST` will be appropriate for the primitive/conserved arrays in `UserWorkInLoop`, primitive arrays in boundary functions, and `ruser_meshblock_data` (assuming it was allocated to be large enough). I think at this point I would need to see the actual code and actual error to tell what's going wrong.

**Reply by tomidakn** [reply]
I agree with Chris. We cannot tell what is actually going wrong without seeing the code and error

**Reply by kyriakosTapinou** [reply]
Thanks both for your comments, they were very helpful and I found the issue in the end it was non-corresponding bounds being used in UserWorkInLoop and in the boundary function.
